pkg/standalone-utils/test/GaugeActions.test.ts

```diff
@@ -0,0 +1,659 @@
+        sharedBeforeEach('', async () => {
+          tokenSender = userSender;
+          tokenRecipient = relayer;
+        });
+
+        itDepositsWithEnoughTokens();
+      });
+
+      context('sender = relayer, recipient = senderUser', () => {
+        sharedBeforeEach('', async () => {
+          await lpToken.connect(userSender).transfer(relayer.address, totalLpTokens);
+          tokenSender = relayer;
+          tokenRecipient = userSender;
+        });
+
+        itDepositsWithEnoughTokens();
+      });
+
+      context('sender = relayer, recipient = relayer', () => {
+        sharedBeforeEach('', async () => {
+          await lpToken.connect(userSender).transfer(relayer.address, totalLpTokens);
+          tokenSender = relayer;
+          tokenRecipient = relayer;
+        });
+
+        itDepositsWithEnoughTokens();
+      });
+    });
+
+    function itDepositsWithEnoughTokens() {
+      context('when depositing some of the tokens', () => {
+        itDepositsWithRefsAndAmounts(totalLpTokens.div(3));
+      });
+
+      context('when depositing all of the available tokens', () => {
+        itDepositsWithRefsAndAmounts(totalLpTokens);
+      });
+
+      context('when depositing 0 tokens', () => {
+        itDepositsWithRefsAndAmounts(bn(0));
+      });
+    }
+
+    function itDepositsWithRefsAndAmounts(amount: BigNumber) {
+      context('when using immediate amounts', () => {
+        itDepositsTokens(amount, amount);
+      });
+
+      context('when using chained references', () => {
+        const reference = toChainedReference(123);
+
+        sharedBeforeEach('set chained reference', async () => {
+          await setChainedReferenceContents(relayer, reference, amount);
+        });
+
+        itDepositsTokens(reference, amount);
+      });
+    }
+
+    function itDepositsTokens(depositAmountOrRef: BigNumber, expectedAmount: BigNumber) {
+      let receipt: ContractReceipt;
+      let tokenSenderAddress: string;
+      let tokenRecipientAddress: string;
+
+      sharedBeforeEach('get addresses', async () => {
+        tokenSenderAddress = TypesConverter.toAddress(tokenSender);
+        tokenRecipientAddress = TypesConverter.toAddress(tokenRecipient);
+      });
+
+      sharedBeforeEach('check initial balances and make deposit', async () => {
+        // Start: sender has all the BPT, no gauge tokens minted, recipient is clean unless it is sender.
+        expect(await lpToken.balanceOf(gauge.address)).to.be.eq(0);
+        expect(await lpToken.balanceOf(tokenSenderAddress)).to.be.eq(totalLpTokens);
+        if (tokenSenderAddress != tokenRecipientAddress) {
+          expect(await lpToken.balanceOf(tokenRecipientAddress)).to.be.eq(0);
+        }
+
+        expect(await gauge.balanceOf(gauge.address)).to.be.eq(0);
+        expect(await gauge.balanceOf(tokenSenderAddress)).to.be.eq(0);
+        expect(await gauge.balanceOf(tokenRecipientAddress)).to.be.eq(0);
```

Comment:
This is fine, but it makes the other tests a bit fragile. I think it'd be better to measure and store the balances _before_ the deposit, and to have the tests work with deltas (i.e. measure balance, assert on `curr - old`)

Reply:
Done; also added a mini fix for `almostEqual` in big number comparisons to support negative numbers with some error.